fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java

```diff
@@ -63,12 +64,55 @@ public static void clearDataSourceContext() {
         contextHolder.remove();
     }
 
+    public static String getAuthToken() {
+        return authTokenContext.get();
+    }
+
     public static void setAuthToken(final String authToken) {
         authTokenContext.set(authToken);
     }
 
-    public static String getAuthToken() {
-        return authTokenContext.get();
+    public static HashMap<BusinessDateType, LocalDate> getBusinessDates() {
+        Assert.notNull(businessDateContext.get(), "Business dates cannot be null!");
+        return businessDateContext.get();
+    }
+
+    public static void setBusinessDates(HashMap<BusinessDateType, LocalDate> dates) {
+        Assert.notNull(dates, "Business dates cannot be null!");
+        businessDateContext.set(dates);
+    }
+
+    public static LocalDate getBusinessDateByType(BusinessDateType businessDateType) {
+        Assert.notNull(businessDateType, "Business date type cannot be null!");
+        LocalDate localDate = getBusinessDates().get(businessDateType);
+        Assert.notNull(localDate, String.format("Business date with type `%s` is not initialised!", businessDateType));
+        return localDate;
+    }
+
+    public static LocalDate getBusinessDate() {
+        BusinessDateType businessDateType = getActionContext().getBusinessDateType();
+        return getBusinessDateByType(businessDateType);
     }
 
+    public static ActionContext getActionContext() {
+        return actionContext.get() == null ? ActionContext.DEFAULT : actionContext.get();
+    }
+
+    public static void setActionContext(ActionContext context) {
+        Assert.notNull(context, "context cannot be null");
+        actionContext.set(context);
+    }
+
+    public static FineractContext syncDown() {
```

Comment:
`syncDown` is a bit confusing for me cause it's not "syncing" anything per se. I'd simply call it something like `ThreadLocalContextUtil#getContext`/`ThreadLocalContextUtil#toContext`